# 01 — Data Cleaning: Orders (2024–present)

**Goal:** pull every order from `trn.dbo.tblOrder` since 2024-01-01, audit it for data-quality issues, and produce a clean CSV (`data/cleaned_orders_2024plus.csv`) that every later analysis notebook can load directly — without re-querying SQL Server or re-doing this cleaning logic.

**Why 2024+ only?** Earlier exploration found 50,109 orders from 2024 onward, out of 106,393 total. Restricting to this window keeps the analysis focused on the recent period relevant to the ~$70k/month recovery goal, and avoids dragging in older POS-system-era data that may follow different conventions.

This notebook works in three phases, each reviewed before moving to the next:
1. **Audit** — connect, pull the data, find every issue, log it (nothing gets dropped yet)
2. **Clean** — apply the agreed-upon exclusions, save the clean CSV
3. **Summarize** — report what was kept, what was cut, and why it matters for later analysis


## Setup

We need four kinds of tools here:
- **`sqlalchemy`** — the standard way for pandas to talk to a SQL database. It manages the connection ("engine") so we can hand SQL text to pandas and get a DataFrame back.
- **`pymssql`** — the actual driver that speaks SQL Server's wire protocol. (We're using this instead of the more common `pyodbc` because `pyodbc` needs Microsoft's ODBC driver installed at the OS level, and that install was blocked by outdated Xcode Command Line Tools on this machine. `pymssql` bundles everything it needs in the Python package itself, so it sidesteps that.)
- **`python-dotenv`** — loads the database password from a local `.env` file instead of us typing it into the notebook. `.env` is in `.gitignore`, so the password never ends up on GitHub.
- **`pandas`** — for everything else.


In [1]:
import os
import re

import pandas as pd
from sqlalchemy import create_engine, text
from dotenv import load_dotenv

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 120)


## A note on privacy

This repo is public. A few cells below print real customer phone numbers (sample values, frequency tables) to make patterns visible directly on GitHub without anyone needing to re-run the notebook. Everywhere a CID gets displayed, it's passed through `mask_cid()` first — area code and last 4 digits stay visible, the exchange is hidden. This only affects what's *printed* in this notebook; the actual data used for analysis (`orders`, `clean_orders`) and the CSV this notebook saves keep full real values, since those aren't committed to git.


In [2]:
def mask_cid(value):
    """Partially mask a CID/phone value for display: area code + last 4 digits visible,
    exchange hidden. Only used at print/display sites -- computation elsewhere in this
    notebook (and the saved CSV) uses the real, unmasked values."""
    if not isinstance(value, str) or len(value) < 13 or value[0] != "(":
        return value
    area = value[1:4]
    last4 = value[9:13]
    return f"({area}) XXX-{last4}"


## Connect to SQL Server

`load_dotenv()` reads the `.env` file in the project root and adds its contents to `os.environ`, so `os.environ["DB_PASSWORD"]` works below without the password ever being written in this notebook.

The connection string format for SQLAlchemy is `dialect+driver://user:password@host:port/database`. Here that's `mssql+pymssql://...` — "mssql" tells SQLAlchemy which SQL dialect to generate, "pymssql" tells it which driver to hand that SQL to.

We connect to the `trn` database specifically (not `master`), since that's where `tblOrder` lives.


In [3]:
load_dotenv()

conn_str = (
    f"mssql+pymssql://{os.environ['DB_USER']}:{os.environ['DB_PASSWORD']}"
    f"@{os.environ['DB_HOST']}:{os.environ['DB_PORT']}/{os.environ['DB_NAME']}"
)
engine = create_engine(conn_str)

# quick smoke test — if this prints without error, the connection works
with engine.connect() as conn:
    print("Connected. Server says:", conn.execute(text("SELECT @@VERSION")).scalar()[:60])


Connected. Server says: Microsoft SQL Server 2022 (RTM-CU25) (KB5081477) - 16.0.4255


## Pull the orders

A quick note on column names: `tblOrder` uses **camelCase** (`checkinTime`, `paidAmount`, `customerID`, etc.), not the PascalCase names (`CheckInTime`, `CID`, ...) used loosely to describe them. SQL Server identifiers aren't case-sensitive by default, but we'll use the real casing from `INFORMATION_SCHEMA.COLUMNS` to avoid confusion later.

We select:
- `ID` — the order's unique identifier (our row identifier for the exclusions log)
- `checkinTime` — when the order started; this is what we filter on for the 2024+ window
- `paidAmount` — the actual revenue figure this whole analysis hinges on
- `paymentType`, `orderType` — for later segmentation (e.g. dine-in vs. takeout, cash vs. card)
- `customerID` — the field described as "CID" / customer phone
- `invoiceStatus`, `kitchenStatus` — order lifecycle codes, useful context for judging whether an odd `paidAmount` is a data error or a legitimately voided/comped order

We filter with `checkinTime >= '2024-01-01'` directly in SQL rather than pulling everything and filtering in pandas — no reason to drag ~56K rows of pre-2024 data over the wire just to discard them.


In [4]:
query = text("""
    SELECT
        ID,
        checkinTime,
        paidAmount,
        paymentType,
        orderType,
        customerID,
        invoiceStatus,
        kitchenStatus
    FROM tblOrder
    WHERE checkinTime >= '2024-01-01'
""")

with engine.connect() as conn:
    orders = pd.read_sql(query, conn)

starting_row_count = len(orders)
print(f"Pulled {starting_row_count:,} orders")
orders.assign(customerID=orders["customerID"].map(mask_cid)).head()


Pulled 50,109 orders


,ID,checkinTime,paidAmount,paymentType,orderType,customerID,invoiceStatus,kitchenStatus
0,10224606DB8B70,2024-01-01 11:27:15,79.19,Cash,1,(631) XXX-9196,1,1
1,00224615F91783,2024-01-02 10:33:19,50.78,Cash,0,(934) XXX-8588,1,1
2,102246255F4830,2024-01-03 11:25:20,8.96,Cash,1,(516) XXX-7881,1,1
3,00124634A01633,2024-01-04 11:21:50,59.04,Cash,0,(347) XXX-7144,1,1
4,10224643C55A41,2024-01-05 10:47:30,20.96,Visa,1,(855) XXX-3333,1,1


## Audit 1: Duplicate order IDs

`ID` should be a primary key — one row per order. If the same `ID` shows up more than once, that's a sign of either a bad join somewhere upstream, a data entry duplicate, or a re-used ID, and we'd want to understand which before deciding whether to keep, merge, or drop those rows.

`duplicated(keep=False)` flags **every** row that shares an `ID` with at least one other row (not just the 2nd, 3rd, etc. occurrence) — that way we can see the full set of duplicates together instead of one row disappearing from view.


In [5]:
dup_id_mask = orders["ID"].duplicated(keep=False)
duplicate_ids = orders[dup_id_mask]

print(f"Duplicate order IDs found: {orders['ID'].duplicated().sum()}")
if len(duplicate_ids) > 0:
    display(duplicate_ids.assign(customerID=duplicate_ids["customerID"].map(mask_cid)).sort_values("ID"))
else:
    print("None — ID is a clean primary key over this date range.")


Duplicate order IDs found: 0
None — ID is a clean primary key over this date range.


## Audit 2: `paidAmount` — null, zero, negative

This is the field any revenue analysis will sum, so it matters most. Three failure modes to check for:
- **null** — no amount recorded at all
- **negative** — would only make sense as a refund/adjustment, and this table doesn't appear to model those explicitly
- **zero** — ambiguous on its own: could be a data error, or a legitimate $0 order (comped meal, cancelled/voided order that never got a real transaction)

For the zero-dollar rows, we cross-tabulate against `invoiceStatus` and `kitchenStatus` rather than just counting them — the *pattern* of those status codes tells us whether the zero-dollar orders look like the same population as normal completed orders (ambiguous, needs a business call) or clearly distinct from them (safer to treat as void/incomplete).


In [6]:
null_paid = orders["paidAmount"].isna()
negative_paid = orders["paidAmount"] < 0
zero_paid = orders["paidAmount"] == 0

print(f"null paidAmount:     {null_paid.sum():,}")
print(f"negative paidAmount: {negative_paid.sum():,}")
print(f"zero paidAmount:     {zero_paid.sum():,}")

print("\nzero-paid orders, broken down by (invoiceStatus, kitchenStatus):")
zero_breakdown = (
    orders[zero_paid]
    .groupby(["invoiceStatus", "kitchenStatus"], dropna=False)
    .size()
    .rename("count")
    .reset_index()
    .sort_values("count", ascending=False)
)
display(zero_breakdown)

print("\nFor comparison, invoiceStatus distribution across ALL 2024+ orders:")
display(orders["invoiceStatus"].value_counts().rename("count"))


null paidAmount:     0
negative paidAmount: 0
zero paidAmount:     406

zero-paid orders, broken down by (invoiceStatus, kitchenStatus):


,invoiceStatus,kitchenStatus,count
1,1,1,366
0,0,0,38
2,2,1,1
3,3,1,1



For comparison, invoiceStatus distribution across ALL 2024+ orders:


invoiceStatus
1    47855
2     2141
3       66
0       39
4        6
5        1
8        1
Name: count, dtype: int64

**🚩 Flagging, not deciding:** `invoiceStatus == 1` is by far the most common status across *all* 2024+ orders (~95% of them), i.e. it's the code for a normally-completed order — yet 366 of the 406 zero-paid rows also carry `invoiceStatus == 1`. That means most zero-dollar orders look, status-wise, identical to normal completed orders. We don't have a lookup table decoding what these status codes mean, so we can't tell from this table alone whether those 366 are comped meals (legitimately $0, should stay in the data) or a data-entry gap (should be excluded from revenue totals). This needs a business call, not a coin flip — it goes into the exclusions log as its own reason category so it can be decided on explicitly rather than silently.

The remaining 40 zero-paid rows (`invoiceStatus` 0, 2, 3) sit outside the normal-order pattern and more plausibly represent voided/incomplete orders.

No null or negative `paidAmount` rows exist in this window — nothing to log for those.


## Audit 3: `customerID` (CID) — coverage baseline

Before looking at *formatting*, let's establish how many rows have a CID at all. We check for both true SQL `NULL` and empty/whitespace-only strings, since a field can be "missing" either way.


In [7]:
has_cid = orders["customerID"].notna() & (orders["customerID"].str.strip() != "")
missing_cid = ~has_cid

print(f"orders with a CID:    {has_cid.sum():,} ({100 * has_cid.mean():.1f}%)")
print(f"orders missing a CID: {missing_cid.sum():,} ({100 * missing_cid.mean():.1f}%)")


orders with a CID:    43,041 (85.9%)
orders missing a CID: 7,068 (14.1%)


## Audit 4: `customerID` formatting

The task description assumed we'd find inconsistent formatting — mixed spacing, dashes, leading zeros stripped. Let's actually check, rather than assume. We'll look at the raw string length distribution first, since that tells us right away whether the POS system is storing this consistently or not.


In [8]:
cid_present = orders.loc[has_cid, "customerID"]

print("Raw string length distribution:")
display(cid_present.str.len().value_counts().sort_index())

print("\nSample values (masked):")
print(cid_present.head(10).map(mask_cid).tolist())


Raw string length distribution:


customerID
20    43041
Name: count, dtype: int64


Sample values (masked):
['(631) XXX-9196', '(934) XXX-8588', '(516) XXX-7881', '(347) XXX-7144', '(855) XXX-3333', '(646) XXX-1584', '(631) XXX-6546', '(631) XXX-1040', '(631) XXX-4722', '(516) XXX-2693']


**Finding:** every CID is stored as a fixed 20-character string. So the "inconsistent spacing/dashes" concern doesn't actually apply here — the POS system enforces one template: `(NPA)NXX-XXXXxEEEEEE`, i.e. a formatted 10-digit phone number followed by a literal `x` and a 6-character extension slot.

The real question is what's *inside* that template. Two things to check next:
1. Does the main 10-digit number ever contain something other than digits (besides the formatting punctuation `()`,`-`)? SQL Server's `nvarchar` won't stop the POS app from writing garbage in there.
2. What's in the extension slot when there's no real extension?

We check this by stripping everything that isn't a digit or underscore, since we've seen `_` show up as a placeholder character in raw samples above.


In [9]:
# the template is "(NPA)NXX-XXXX" (13 chars) + "x" (1 char) + 6-char extension slot = 20 chars
main_number = cid_present.str[0:13]
extension = cid_present.str[14:20]

# how many digits are actually present in the main number (vs. '_' placeholders)?
main_digit_count = main_number.str.replace(r"\D", "", regex=True).str.len()

print("Digit count within the main number portion (should be 10 if fully captured):")
display(main_digit_count.value_counts().sort_index())

incomplete_main = main_digit_count < 10
print(f"\nCIDs with an incomplete main number (underscore placeholders): {incomplete_main.sum():,}")
print("Examples (masked):")
print(main_number[incomplete_main].head(5).map(mask_cid).tolist())

has_extension = extension.str.contains(r"\d", regex=True)
print(f"\nCIDs with a real (non-blank) extension: {has_extension.sum():,}")


Digit count within the main number portion (should be 10 if fully captured):


customerID
4         1
5         1
6         1
7         1
8         2
9        11
10    43024
Name: count, dtype: int64


CIDs with an incomplete main number (underscore placeholders): 17
Examples (masked):
['(631) XXX-196_', '(631) XXX-10__', '(518) XXX-828_', '(516) XXX-6___', '(516) XXX-882_']

CIDs with a real (non-blank) extension: 75


## Audit 5: obviously invalid / test CIDs

Now the important check: does the same CID value repeat *far* more than any one customer plausibly would? A phone number appearing on 2,000+ orders isn't 2,000 loyal visits from one person — it's almost certainly a **house number**: what the front-of-house staff types in when a real customer's phone wasn't captured (walk-ins, phone orders where the caller ID system defaulted, etc).

We check by looking at the concentration of the top values relative to total CID-bearing rows.


In [10]:
value_counts = cid_present.value_counts()

print(f"Distinct CID values: {cid_present.nunique():,} across {len(cid_present):,} rows")
print("\nTop 15 most frequent CID values (masked):")
top15 = value_counts.head(15)
top15_pct = (top15 / len(cid_present) * 100).round(2)
top15_display = pd.DataFrame({"count": top15, "% of all CID rows": top15_pct})
top15_display.index = top15_display.index.map(mask_cid)
display(top15_display)

print(f"\nThose top 15 values alone account for {100 * top15.sum() / len(cid_present):.1f}% of all CID-bearing rows.")


Distinct CID values: 7,920 across 43,041 rows

Top 15 most frequent CID values (masked):


,count,% of all CID rows
customerID,,
(631) XXX-1044,2208,5.13
(631) XXX-1045,1071,2.49
(631) XXX-1040,848,1.97
(631) XXX-1048,741,1.72
(631) XXX-1042,582,1.35
(631) XXX-1049,421,0.98
(631) XXX-1046,275,0.64
(631) XXX-1047,239,0.56
(855) XXX-1045,171,0.40



Those top 15 values alone account for 17.2% of all CID-bearing rows.


**🚩 Correction after review:** Tang's actual phone number is `(631) XXX-9196` (masked here since this notebook is public) — it shows up in the data too (153 times), but it's *not* this cluster. So `973-10XX` is not the restaurant's own line.

It's still very unlikely to be genuine repeat customers, though — the cluster also appears under a toll-free area code, which an individual customer wouldn't have personally, and the same `973-10XX` exchange/line-number block repeats across multiple different area codes, which a real single phone line never would (see the masked frequency table below). That combination points to a **system-generated default** — most plausibly a delivery platform (DoorDash / Grubhub / UberEats) injecting a masked/proxy number, or a hardcoded POS software default — rather than Tang's own placeholder. **The exact source is unconfirmed from this data alone**, so we'll flag it as such rather than asserting it.

As a bonus check (not blocking): if this cluster correlates strongly with a specific `orderType` or `paymentType` code, that's supporting evidence for a systemic (not staff-behavior) origin, even without a lookup table to decode what the codes mean.


In [11]:
# candidate "systemic default" pattern — not Tang's own number (confirmed separately, masked above),
# but 973-10XX repeating under multiple area codes (incl. toll-free 855) rules out a genuine single customer
systemic_default_mask = orders["customerID"].fillna("").str.match(r"^\(\d{3}\)973-10\d{2}x")

# all-zero placeholder in the main number
main_digits_full = orders["customerID"].fillna("").str[0:13].str.replace(r"\D", "", regex=True)
all_zero_mask = main_digits_full == "0000000000"

print(f"Rows matching the systemic-default pattern (973-10XX): {systemic_default_mask.sum():,} "
      f"({100 * systemic_default_mask.sum() / starting_row_count:.1f}% of all 2024+ orders)")
print(f"Rows with an all-zero placeholder number:                {all_zero_mask.sum():,}")


Rows matching the systemic-default pattern (973-10XX): 7,032 (14.0% of all 2024+ orders)
Rows with an all-zero placeholder number:                1


### Bonus: does the `973-10XX` cluster correlate with `orderType` / `paymentType`?

We don't have a lookup table decoding what each `orderType`/`paymentType` code means, so this can't *prove* a source — but if the cluster's distribution across these codes looks nothing like the overall order population, that's evidence it comes from a specific, systemic channel rather than scattered staff behavior across all order types.


In [12]:
print("orderType distribution — 973-10XX cluster:")
display(orders.loc[systemic_default_mask, "orderType"].value_counts(normalize=True).round(3))

print("\norderType distribution — ALL 2024+ orders (for comparison):")
display(orders["orderType"].value_counts(normalize=True).round(3))

print("\npaymentType distribution — 973-10XX cluster:")
display(orders.loc[systemic_default_mask, "paymentType"].value_counts(normalize=True).round(3))

print("\npaymentType distribution — ALL 2024+ orders (for comparison):")
display(orders["paymentType"].value_counts(normalize=True).round(3))


orderType distribution — 973-10XX cluster:


orderType
1    1.0
2    0.0
Name: proportion, dtype: float64


orderType distribution — ALL 2024+ orders (for comparison):


orderType
1    0.564
0    0.300
2    0.136
4    0.000
3    0.000
Name: proportion, dtype: float64


paymentType distribution — 973-10XX cluster:


paymentType
Visa         0.513
Cash         0.476
Master       0.007
Discover     0.003
             0.000
Check        0.000
Othercard    0.000
Name: proportion, dtype: float64


paymentType distribution — ALL 2024+ orders (for comparison):


paymentType
Cash         0.700
Visa         0.284
Master       0.011
Discover     0.002
             0.001
VOID         0.001
Diner        0.000
Othercard    0.000
Amex         0.000
BarComp      0.000
Check        0.000
Name: proportion, dtype: float64

**Finding:** the `973-10XX` cluster is **100% a single `orderType` code** (the one that's the plurality-but-not-majority — 56.4% — of all orders), whereas the overall population splits across 3 different `orderType` codes. It also skews notably toward card payment (51.3% Visa vs. 28.4% baseline) and away from cash (47.6% vs. 70% baseline).

That 100%-concentration-in-one-code result is a real signal: this cluster isn't scattered across order types the way genuine walk-in "staff didn't ask for a number" behavior would be — it's tied to one specific channel. That's consistent with the delivery-platform or POS-default hypothesis, though without a code-to-label lookup table we can't say *which* one. Treat this as supporting evidence, not confirmation.


## Building the exclusions log

One important judgment call before we log anything: **a bad CID doesn't mean the order itself is bad.** The order still represents a real transaction with a real `paidAmount` — the phone number just can't be trusted for customer-level analysis (repeat-visit counts, customer lifetime value, etc). So this log distinguishes two kinds of reasons via a `proposed_action` column:

- `exclude_row` — the row itself is unreliable as revenue data and should probably be dropped or set aside from revenue totals (duplicate IDs, zero-paid orders)
- `flag_cid_only` — the order stays, but its `customerID` shouldn't be trusted for customer-grouping analysis (systemic-default numbers, all-zero placeholders, incomplete numbers)

Nothing is actually dropped in this cell — we're just building the record. Phase 2 will apply whatever we agree on.


In [13]:
exclusion_frames = []

# --- exclude_row candidates ---

if dup_id_mask.sum() > 0:
    exclusion_frames.append(pd.DataFrame({
        "ID": orders.loc[dup_id_mask, "ID"],
        "reason": "duplicate_order_id",
        "proposed_action": "exclude_row",
    }))

# zero-paid, ambiguous subgroup (same invoiceStatus as normal completed orders)
zero_ambiguous = zero_paid & (orders["invoiceStatus"] == 1)
exclusion_frames.append(pd.DataFrame({
    "ID": orders.loc[zero_ambiguous, "ID"],
    "reason": "zero_paid_amount_ambiguous_status1",
    "proposed_action": "exclude_row",
}))

# zero-paid, non-standard status (more plausibly void/incomplete)
zero_other = zero_paid & (orders["invoiceStatus"] != 1)
exclusion_frames.append(pd.DataFrame({
    "ID": orders.loc[zero_other, "ID"],
    "reason": "zero_paid_amount_nonstandard_status",
    "proposed_action": "exclude_row",
}))

# --- flag_cid_only candidates (order itself is kept) ---

# NOT Tang's own number (confirmed separately, masked above).
# 973-10XX repeats under multiple area codes incl. toll-free 855, and is 100% concentrated in
# one orderType code -- points to a systemic default (delivery platform proxy number or POS
# default) rather than real customers. Source unconfirmed -- see markdown above.
exclusion_frames.append(pd.DataFrame({
    "ID": orders.loc[systemic_default_mask, "ID"],
    "reason": "cid_matches_systemic_default_pattern_973_10xx_source_unconfirmed",
    "proposed_action": "flag_cid_only",
}))

exclusion_frames.append(pd.DataFrame({
    "ID": orders.loc[all_zero_mask, "ID"],
    "reason": "cid_all_zero_placeholder",
    "proposed_action": "flag_cid_only",
}))

incomplete_full = has_cid & (main_digits_full.str.len() < 10)
exclusion_frames.append(pd.DataFrame({
    "ID": orders.loc[incomplete_full, "ID"],
    "reason": "cid_incomplete_digits",
    "proposed_action": "flag_cid_only",
}))

exclusions_log = pd.concat(exclusion_frames, ignore_index=True)

# a single order can trip more than one check (e.g. house number AND zero-paid) —
# keep every (ID, reason) pair rather than collapsing, since each is independently useful information
exclusions_log = exclusions_log.drop_duplicates()

print(f"Exclusions log: {len(exclusions_log):,} entries covering {exclusions_log['ID'].nunique():,} distinct orders")
print("\nBy reason:")
display(exclusions_log["reason"].value_counts())
print("\nBy proposed action:")
display(exclusions_log["proposed_action"].value_counts())


Exclusions log: 7,456 entries covering 7,455 distinct orders

By reason:


reason
cid_matches_systemic_default_pattern_973_10xx_source_unconfirmed    7032
zero_paid_amount_ambiguous_status1                                   366
zero_paid_amount_nonstandard_status                                   40
cid_incomplete_digits                                                 17
cid_all_zero_placeholder                                               1
Name: count, dtype: int64


By proposed action:


proposed_action
flag_cid_only    7050
exclude_row       406
Name: count, dtype: int64

## Phase 1 summary

Note: rows missing a CID entirely (~14%, quantified above) aren't in this log — that's a documented absence, not an anomaly requiring a judgment call. It'll be called out explicitly in the Phase 3 write-up instead.


In [14]:
rows_flagged_for_exclusion = exclusions_log.loc[
    exclusions_log["proposed_action"] == "exclude_row", "ID"
].nunique()
systemic_default_pct_of_all = 100 * systemic_default_mask.sum() / starting_row_count

print("=" * 60)
print("PHASE 1 COMPLETE — awaiting review before Phase 2")
print("=" * 60)
print(f"Starting rows (checkinTime >= 2024-01-01): {starting_row_count:,}")
print(f"Distinct orders flagged as exclude_row:     {rows_flagged_for_exclusion:,}")
print(f"Distinct orders flagged flag_cid_only:      "
      f"{exclusions_log.loc[exclusions_log['proposed_action'] == 'flag_cid_only', 'ID'].nunique():,}")
print(f"CID missing entirely:                       {missing_cid.sum():,} ({100 * missing_cid.mean():.1f}%)")
print()
print("Resolved: 973-10XX is NOT Tang's own line (that's (631) XXX-9196, confirmed separately).")
print(f"  It's flagged as flag_cid_only ({systemic_default_pct_of_all:.1f}% of all orders, "
      f"{systemic_default_mask.sum():,} rows) — likely a systemic default (delivery platform or")
print("  POS software), source unconfirmed. 100% concentration in one orderType code supports this.")
print()
print("Still open before Phase 2 applies anything:")
print("  Should invoiceStatus==1 zero-paid orders (366 rows) be excluded from revenue,")
print("  or are these legitimate $0 comps that should stay?")


PHASE 1 COMPLETE — awaiting review before Phase 2
Starting rows (checkinTime >= 2024-01-01): 50,109
Distinct orders flagged as exclude_row:     406
Distinct orders flagged flag_cid_only:      7,050
CID missing entirely:                       7,068 (14.1%)

Resolved: 973-10XX is NOT Tang's own line (that's (631) XXX-9196, confirmed separately).
  It's flagged as flag_cid_only (14.0% of all orders, 7,032 rows) — likely a systemic default (delivery platform or
  POS software), source unconfirmed. 100% concentration in one orderType code supports this.

Still open before Phase 2 applies anything:
  Should invoiceStatus==1 zero-paid orders (366 rows) be excluded from revenue,
  or are these legitimate $0 comps that should stay?


# Phase 2 — Apply exclusions, build the clean dataset

## Reclassifying the 366 ambiguous zero-paid orders

Decision: **don't hard-drop the 366 `invoiceStatus==1` zero-paid orders.** A $0 comped order still represents a real customer visit — dropping the row entirely would understate order volume and visit frequency for any future loyalty/frequency analysis. But it *would* distort revenue metrics (total revenue, average ticket size) if left in unmarked, since $0 pulls those numbers down without representing real spend.

So instead of `exclude_row`, this group gets a third action category: **`flag_zero_paid`** — the row stays in the clean dataset, but carries a flag that later notebooks can use to exclude it from revenue-specific calculations while still counting it in volume/frequency ones.

The other 40 zero-paid rows (`invoiceStatus` 0/2/3 — the non-standard-status group that looks like voids/incomplete orders rather than completed visits) are **not** part of this change and remain `exclude_row`, since a voided order isn't confidently a completed visit the way a same-status $0 comp is. Flagging this distinction explicitly here in case you want that revisited too — it wasn't part of what you decided on, just carried over from Phase 1.


In [15]:
# reclassify: zero_paid_amount_ambiguous_status1 moves from exclude_row -> flag_zero_paid
reclassify_mask = exclusions_log["reason"] == "zero_paid_amount_ambiguous_status1"
exclusions_log.loc[reclassify_mask, "proposed_action"] = "flag_zero_paid"

print("Updated exclusions log — by proposed_action:")
display(exclusions_log["proposed_action"].value_counts())


Updated exclusions log — by proposed_action:


proposed_action
flag_cid_only     7050
flag_zero_paid     366
exclude_row         40
Name: count, dtype: int64

## Build the clean dataframe

Three things happen here:
1. **Drop** the rows still marked `exclude_row` (duplicate IDs + the 40 non-standard-status zero-paid rows) — these are the only rows judged unreliable enough to remove entirely.
2. **Attach `excluded_from_revenue`** (bool) — `True` for the 366 reclassified zero-paid rows. Any future revenue or average-ticket-size calculation should filter on `~excluded_from_revenue`; order-count/visit-frequency analyses should ignore this column and use every row.
3. **Attach `cid_flag`** — the specific reason a CID shouldn't be trusted (`cid_matches_systemic_default_pattern_973_10xx_source_unconfirmed`, `cid_all_zero_placeholder`, or `cid_incomplete_digits`), and **`cid_usable`** (bool) as a one-column shortcut: `True` only when a CID is present *and* untainted by any of those flags.


In [16]:
# 1. drop the true row-level exclusions
excluded_row_ids = exclusions_log.loc[exclusions_log["proposed_action"] == "exclude_row", "ID"]
clean_orders = orders[~orders["ID"].isin(excluded_row_ids)].copy()

# 2. flag rows to exclude from revenue-specific calcs (but keep for volume/frequency)
zero_paid_flag_ids = exclusions_log.loc[exclusions_log["proposed_action"] == "flag_zero_paid", "ID"]
clean_orders["excluded_from_revenue"] = clean_orders["ID"].isin(zero_paid_flag_ids)

# 3. attach the CID trust flag — each order has at most one cid_flag reason (categories are
# mutually exclusive by construction), so this is a safe 1-to-1 merge
cid_flags = (
    exclusions_log.loc[exclusions_log["proposed_action"] == "flag_cid_only", ["ID", "reason"]]
    .drop_duplicates(subset="ID")
    .rename(columns={"reason": "cid_flag"})
)
clean_orders = clean_orders.merge(cid_flags, on="ID", how="left")

cid_present_clean = clean_orders["customerID"].notna() & (clean_orders["customerID"].str.strip() != "")
clean_orders["cid_usable"] = cid_present_clean & clean_orders["cid_flag"].isna()

print(f"Rows dropped (exclude_row):     {starting_row_count - len(clean_orders):,}")
print(f"Rows in clean dataset:          {len(clean_orders):,}")
print(f"  excluded_from_revenue=True:   {clean_orders['excluded_from_revenue'].sum():,}")
print(f"  cid_usable=True:              {clean_orders['cid_usable'].sum():,} "
      f"({100 * clean_orders['cid_usable'].mean():.1f}%)")
clean_orders.assign(customerID=clean_orders["customerID"].map(mask_cid)).head()


Rows dropped (exclude_row):     40
Rows in clean dataset:          50,069
  excluded_from_revenue=True:   366
  cid_usable=True:              35,966 (71.8%)


,ID,checkinTime,paidAmount,paymentType,orderType,customerID,invoiceStatus,kitchenStatus,excluded_from_revenue,cid_flag,cid_usable
0,10224606DB8B70,2024-01-01 11:27:15,79.19,Cash,1,(631) XXX-9196,1,1,False,NaN,True
1,00224615F91783,2024-01-02 10:33:19,50.78,Cash,0,(934) XXX-8588,1,1,False,NaN,True
2,102246255F4830,2024-01-03 11:25:20,8.96,Cash,1,(516) XXX-7881,1,1,False,NaN,True
3,00124634A01633,2024-01-04 11:21:50,59.04,Cash,0,(347) XXX-7144,1,1,False,NaN,True
4,10224643C55A41,2024-01-05 10:47:30,20.96,Visa,1,(855) XXX-3333,1,1,False,NaN,True


## Save the clean CSV

Saving to `data/cleaned_orders_2024plus.csv` so later notebooks can `pd.read_csv()` this directly instead of re-connecting to SQL Server and re-running this whole audit. `data/` is in `.gitignore` — this file (and the raw phone-number-adjacent data it contains) never gets committed to GitHub.


In [17]:
os.makedirs("../data", exist_ok=True)
clean_orders.to_csv("../data/cleaned_orders_2024plus.csv", index=False)

# also save the exclusions log itself, for auditability — anyone reviewing this analysis
# should be able to see exactly which orders were touched and why
exclusions_log.to_csv("../data/exclusions_log_2024plus.csv", index=False)

print(f"Saved {len(clean_orders):,} rows to data/cleaned_orders_2024plus.csv")
print(f"Saved {len(exclusions_log):,} rows to data/exclusions_log_2024plus.csv")


Saved 50,069 rows to data/cleaned_orders_2024plus.csv
Saved 7,456 rows to data/exclusions_log_2024plus.csv


In [18]:
print("=" * 60)
print("PHASE 2 COMPLETE — awaiting review before Phase 3")
print("=" * 60)
print(f"Starting rows:              {starting_row_count:,}")
print(f"Rows dropped (exclude_row): {starting_row_count - len(clean_orders):,}")
print(f"Clean dataset rows:         {len(clean_orders):,} "
      f"({100 * len(clean_orders) / starting_row_count:.1f}% of starting rows)")
print(f"  of which excluded_from_revenue=True: {clean_orders['excluded_from_revenue'].sum():,}")
print(f"  of which cid_usable=True:            {clean_orders['cid_usable'].sum():,} "
      f"({100 * clean_orders['cid_usable'].mean():.1f}%)")
print()
print("Saved: data/cleaned_orders_2024plus.csv, data/exclusions_log_2024plus.csv")


PHASE 2 COMPLETE — awaiting review before Phase 3
Starting rows:              50,109
Rows dropped (exclude_row): 40
Clean dataset rows:         50,069 (99.9% of starting rows)
  of which excluded_from_revenue=True: 366
  of which cid_usable=True:            35,966 (71.8%)

Saved: data/cleaned_orders_2024plus.csv, data/exclusions_log_2024plus.csv


# Phase 3 — Summary

This is the recap any future notebook (or a human reviewing this analysis) should read first: what came in, what got cut, what got flagged, and what that means for what you can and can't trust this data to answer.


In [19]:
missing_cid_clean = clean_orders["customerID"].isna() | (clean_orders["customerID"].str.strip() == "")

summary_rows = [
    ("Starting rows (checkinTime >= 2024-01-01)", starting_row_count, None),
    ("— dropped: duplicate_order_id", 0, "exclude_row"),
    ("— dropped: zero_paid_amount_nonstandard_status", 40, "exclude_row"),
    ("Clean dataset (rows retained)", len(clean_orders), None),
    ("  of which excluded_from_revenue (zero-paid comps)", int(clean_orders["excluded_from_revenue"].sum()), "flag_zero_paid"),
    ("  of which cid missing entirely", int(missing_cid_clean.sum()), "no reason logged — documented absence"),
    ("  of which cid_flag: systemic_default_973_10xx", int((clean_orders["cid_flag"] == "cid_matches_systemic_default_pattern_973_10xx_source_unconfirmed").sum()), "flag_cid_only"),
    ("  of which cid_flag: incomplete_digits", int((clean_orders["cid_flag"] == "cid_incomplete_digits").sum()), "flag_cid_only"),
    ("  of which cid_flag: all_zero_placeholder", int((clean_orders["cid_flag"] == "cid_all_zero_placeholder").sum()), "flag_cid_only"),
    ("  of which cid_usable=True", int(clean_orders["cid_usable"].sum()), None),
]

for label, count, note in summary_rows:
    pct = f"{100 * count / starting_row_count:.1f}%"
    note_str = f"  [{note}]" if note else ""
    print(f"{label:<55} {count:>7,}  ({pct:>6}){note_str}")

print()
print(f"% of starting data retained as rows: {100 * len(clean_orders) / starting_row_count:.1f}%")
print(f"% of clean rows with a trustworthy customer ID:  {100 * clean_orders['cid_usable'].mean():.1f}%")


Starting rows (checkinTime >= 2024-01-01)                50,109  (100.0%)
— dropped: duplicate_order_id                                 0  (  0.0%)  [exclude_row]
— dropped: zero_paid_amount_nonstandard_status               40  (  0.1%)  [exclude_row]
Clean dataset (rows retained)                            50,069  ( 99.9%)
  of which excluded_from_revenue (zero-paid comps)          366  (  0.7%)  [flag_zero_paid]
  of which cid missing entirely                           7,054  ( 14.1%)  [no reason logged — documented absence]
  of which cid_flag: systemic_default_973_10xx            7,031  ( 14.0%)  [flag_cid_only]
  of which cid_flag: incomplete_digits                       17  (  0.0%)  [flag_cid_only]
  of which cid_flag: all_zero_placeholder                     1  (  0.0%)  [flag_cid_only]
  of which cid_usable=True                               35,966  ( 71.8%)

% of starting data retained as rows: 99.9%
% of clean rows with a trustworthy customer ID:  71.8%


## What this means going forward

**Row-level cleanliness is excellent.** 50,069 of 50,109 orders (99.9%) survive as revenue/order records. Only 40 rows were dropped outright, and those were genuinely ambiguous (zero-paid orders with non-standard status codes, not just any $0 order). The 366 zero-paid orders under normal status stay in the data, tagged `excluded_from_revenue=True` — remember to filter on that column for any revenue or average-ticket-size number, but *not* for order-count or visit-frequency numbers.

**🚩 The customer-identification gap is bigger than the original ~14% estimate — it's closer to 28%.** That original number only counted orders with *no* CID at all (7,054 rows, 14.1%). But on top of that, another 7,049 rows (14.1%) *have* a CID that's flagged as unreliable — almost entirely the `973-10XX` systemic-default cluster. Put together, **only 71.8% of clean orders (`cid_usable=True`) have a customer identifier worth trusting.**

Concretely, for any future analysis:
- **Revenue / order-volume / menu-mix analysis:** unaffected — use the full clean dataset (filtering `excluded_from_revenue` where relevant). CID quality doesn't matter here.
- **Customer-level analysis** (repeat-visit rate, customer lifetime value, loyalty segmentation, "how many unique customers"): **filter to `cid_usable=True` first.** Skipping that filter will silently lump ~28% of orders into a handful of fake "customers" (the systemic-default numbers most of all — one single value alone accounts for 5% of all CID-bearing rows), which would badly distort repeat-visit and concentration metrics.
- Even within `cid_usable=True`, treat customer-level conclusions as describing roughly 72% of order volume, not the full picture — worth stating as a caveat in any write-up that leans on customer grouping.
- The source of the `973-10XX` cluster is still unconfirmed (delivery platform proxy number vs. POS default is a reasonable guess, not a fact). If that's worth nailing down before customer-analysis work starts, the natural next step is checking with Tang's POS vendor or looking at whether `orderType`'s specific code (the one this cluster is 100% concentrated in) is documented anywhere as "phone/delivery order."
